lib/compile/formats.js

```diff
@@ -33,7 +33,8 @@ formats.fast = {
   // email (sources from jsen validator):
   // http://stackoverflow.com/questions/201323/using-a-regular-expression-to-validate-an-email-address#answer-8829363
   // http://www.w3.org/TR/html5/forms.html#valid-e-mail-address (search for 'willful violation')
-  email: /^[a-z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?(?:\.[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)*$/i,
+  // Modified to allow '_' in the domain
+  email: /^[a-z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-z0-9_](?:[a-z0-9-_]{0,61}[a-z0-9_])?(?:\.[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)*$/i,
   hostname: HOSTNAME,
```

Comment:
I think RFC 5321 is the current version of the SMTP standard.
Section 4.1.2 describes the structure of email addresses (among other things).
Here's a relevant section

```
   To promote interoperability and consistent with long-standing
   guidance about conservative use of the DNS in naming and applications
   (e.g., see Section 2.3.1 of the base DNS document, RFC 1035 [2]),
   characters outside the set of alphabetic characters, digits, and
   hyphen MUST NOT appear in domain name labels for SMTP clients or
   servers.  In particular, the underscore character is not permitted.
   SMTP servers that receive a command in which invalid character codes
   have been employed, and for which there are no other reasons for
   rejection, MUST reject that command with a 501 response (this rule,
   like others, could be overridden by appropriate SMTP extensions).
```
So they explicitly prohibit the use of underscore's even in cases where DNS would actually accept them.

Also if you look at how domain names are defined:
```
   Domain         = sub-domain *("." sub-domain)
   sub-domain     = Let-dig [Ldh-str]
   Let-dig        = ALPHA / DIGIT
   Ldh-str        = *( ALPHA / DIGIT / "-" ) Let-dig
```
A domain name cannot begin with a hypen, only a letter or digit

I think the current regex is actually a good representation of the RFC.

I have seen email addresses with underscores in subdomains which seem to work, but if the RFC explicitly rules them out, I don't think it's worth putting it in the regex. Servers that accept them must be considered faulty, and users that have these kind of email addresses will bump into problems sooner rather than later.

In short, after doing more research, I think it's probably best to close my PR and leave the regex as is.